lib/saluki-components/src/encoders/datadog/logs/mod.rs

```diff
@@ -0,0 +1,287 @@
+                    let http_payload = HttpPayload::new(payload_meta, request);
+                    let payload = Payload::Http(http_payload);
+                    dispatcher.dispatch(payload).await?;
+                }
+                Err(e) => error!(error = %e, "Failed to build Datadog logs payload. Continuing..."),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct LogsEndpointEncoder {
+    ddsource: String,
+    tags_deduplicator: ReusableDeduplicator<Tag>,
+}
+
+impl LogsEndpointEncoder {
+    fn new(ddsource: String) -> Self {
+        Self {
+            ddsource,
+            tags_deduplicator: ReusableDeduplicator::new(),
+        }
+    }
+
+    fn status_to_str(status: LogStatus) -> &'static str {
+        match status {
+            LogStatus::Trace => "trace",
+            LogStatus::Emergency => "emerg",
+            LogStatus::Alert => "alert",
+            LogStatus::Fatal => "crit",
+            LogStatus::Error => "err",
+            LogStatus::Warning => "warn",
+            LogStatus::Notice => "notice",
+            LogStatus::Info => "info",
+            LogStatus::Debug => "debug",
+        }
+    }
+
+    // TODO: add source for logs
+
+    fn build_agent_json(&mut self, log: &Log) -> JsonValue {
+        let mut obj = JsonMap::new();
+
+        // Required-ish envelope fields
+        obj.insert("message".to_string(), JsonValue::String(log.message().to_string()));
+
+        if let Some(status) = log.status() {
+            obj.insert(
+                "status".to_string(),
+                JsonValue::String(Self::status_to_str(status).to_string()),
+            );
+        }
+
+        if !log.hostname().is_empty() {
+            obj.insert("hostname".to_string(), JsonValue::String(log.hostname().to_string()));
+        }
+        if !log.service().is_empty() {
+            obj.insert("service".to_string(), JsonValue::String(log.service().to_string()));
+        }
+
+        obj.insert("ddsource".to_string(), JsonValue::String(self.ddsource.clone()));
+
+        // ddtags: comma-separated, deduplicated
+        let tags_iter = self.tags_deduplicator.deduplicated(log.tags().into_iter());
+        let tags_vec: Vec<&str> = tags_iter.map(|t| t.as_str()).collect();
+        if !tags_vec.is_empty() {
+            obj.insert("ddtags".to_string(), JsonValue::String(tags_vec.join(",")));
+        }
+
+        // Copy additional properties flattened into the envelope
+        for (k, v) in log.additional_properties() {
+            obj.insert(k.clone(), v.clone());
+        }
+
+        // If we have "otel.timestamp" as a numeric string, add a numeric "timestamp" field (nanoseconds since epoch).
+        if let Some(JsonValue::String(ns)) = obj.get("otel.timestamp") {
+            if let Ok(parsed) = ns.parse::<i64>() {
+                obj.insert("timestamp".to_string(), JsonValue::from(parsed));
+            }
+        }
```

Comment:
Should we just have a `timestamp` field on `Log` itself? Do we have that for our internal log representation in the Agent?

We should try to be generic here where possible, and handle translations (like using `otel.timestamp` as the log timestamp) as far upstream as possible, ideally in the respective source.